qcodes/dataset/dond/do_nd.py

```diff
@@ -260,44 +678,39 @@ def dond(
 
     try:
         with _catch_interrupts() as interrupted, ExitStack() as stack, params_meas_caller as call_params_meas:
-            datasavers = [stack.enter_context(measure.run()) for measure in meas_list]
+            datasavers = [
+                stack.enter_context(group.measurement_cxt.run())
+                for group in measurements.groups
+            ]
             additional_setpoints_data = process_params_meas(additional_setpoints)
-            previous_setpoints = np.empty(len(sweep_instances))
-            for setpoints in tqdm(sweeper.nested_setpoints, disable=not show_progress):
-
-                active_actions, delays = _select_active_actions_delays(
-                    sweeper.post_actions,
-                    sweeper.post_delays,
-                    setpoints,
-                    previous_setpoints,
-                )
-                previous_setpoints = setpoints
-
-                param_set_list = []
-                for setpoint_param, setpoint, action, delay in zip(
-                    sweeper.params_set,
-                    setpoints,
-                    active_actions,
-                    delays,
-                ):
-                    _conditional_parameter_set(setpoint_param, setpoint)
-                    param_set_list.append((setpoint_param, setpoint))
-                    for act in action:
-                        act()
-                    time.sleep(delay)
+            # _Sweeper is not considered an Iterable since it does not
+            # implement __iter__. However, it does implement __getitem__
+            # and is therefor safe to iterate over.
+            # https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable
+            # https://github.com/python/cpython/issues/86992#issuecomment-1093897307
+            for set_events in tqdm(sweeper, disable=not show_progress):  # type: ignore[call-overload]
+                results: dict[ParameterBase, Any] = {}
+                for set_event in set_events:
+                    if set_event.should_set:
+                        set_event.parameter(set_event.new_value)
+                        for act in set_event.actions:
+                            act()
+                        time.sleep(set_event.delay)
```

Comment:
same for the delay - in the original code it was executed even if the parameter is not set. again, i understand that there's probably no point in sleeping if parameter value was not changed, but that's what the original behavior does and perhaps we don't want to fix it in this PR?